File: bjst/trainer.py
import random
from abc import ABC, abstractmethod
from .strategy import StrategyChart
from .ui import (display_session_header, display_hand, get_user_action,
                 display_feedback, display_dealer_groups, display_hand_types)


class TrainingSession(ABC):
    """Base class for all training session types."""

    def __init__(self, difficulty='normal'):
        self.difficulty = difficulty
        self.strategy = StrategyChart()
        self.correct_count = 0
        self.total_count = 0
        self.session_stats = {}

    @property
    @abstractmethod
    def mode_name(self):
        """Return the mode name for display purposes."""

    @property
    @abstractmethod
    def max_questions(self):
        """Return the maximum number of questions for this session type."""

    @abstractmethod
    def generate_scenario(self):
        """Generate a scenario for this training mode."""

    def _generate_hand_cards(self, hand_type, player_total):
        """Helper method to generate card representation for a hand."""
        if hand_type == 'pair':
            return [player_total, player_total]
        if hand_type == 'soft':
            other_card = player_total - 11
            return [11, other_card]
        # hard
        if player_total <= 11:
            return [player_total]
        # Generate two valid cards (2-10) that sum to player_total
        first_card = random.randint(2, min(10, player_total - 2))
        second_card = player_total - first_card

        # If second card would be > 10, we need more cards
        if second_card > 10:
            # For totals > 20, generate 3+ cards
            cards = [first_card]
            remaining = player_total - first_card

            while remaining > 10:
                # Take a card between 2 and min(10, remaining-2) to ensure we can finish
                max_card = min(10, remaining - 2)
                if max_card < 2:
                    break
                card = random.randint(2, max_card)
                cards.append(card)
                remaining -= card

            if remaining >= 2:
                cards.append(remaining)
            return cards
        if second_card < 2:
            # If second card would be < 2, just use single card
            return [player_total]
        return [first_card, second_card]

    def check_answer(self, user_action, correct_action):
        """Check if user's action matches the correct action."""
        # Handle split variations
        if user_action == 'P':
            user_action = 'Y'
        return user_action == correct_action

    def show_feedback(self, scenario, user_action, correct_action):
        """Display feedback for the user's answer."""
        hand_type, _, player_total, dealer_card = scenario
        correct = self.check_answer(user_action, correct_action)
        explanation = self.strategy.get_explanation(
            hand_type, player_total, dealer_card)
        response = display_feedback(
            correct, user_action, correct_action, explanation)
        return correct, response

    def setup_session(self):
        """Setup the session. Override in subclasses if additional setup is needed."""
        return None  # Return submode if applicable

    def run(self, stats):
        """Run the training session."""
        display_session_header(self.mode_name)

        if hasattr(self, '_requires_submode'):
            submode = self.setup_session()  # pylint: disable=assignment-from-none
            if submode is None:
                return  # User cancelled selection

        question_count = 0
        while question_count < self.max_questions:
            scenario = self.generate_scenario()
            hand_type, player_cards, player_total, dealer_card = scenario

            display_hand(player_cards, dealer_card, hand_type, player_total)

            user_action = get_user_action()
            if user_action is None:  # User quit
                break

            correct_action = self.strategy.get_correct_action(
                hand_type, player_total, dealer_card)
            correct, response = self.show_feedback(
                scenario, user_action, correct_action)

            # Record statistics
            dealer_strength = stats.get_dealer_strength(dealer_card)
            stats.record_attempt(hand_type, dealer_strength, correct)

            question_count += 1

            if correct:
                self.correct_count += 1
            self.total_count += 1

            # Check if user wants to quit
            if response == 'quit':
                break

        # Show session summary
        if self.total_count > 0:
            accuracy = (self.correct_count / self.total_count) * 100
            print(f"\nSession complete! Final score: {self.correct_count}/"
                  f"{self.total_count} ({accuracy:.1f}%)")


class RandomTrainingSession(TrainingSession):
    """Random practice session with all hand types and dealer cards."""

    @property
    def mode_name(self):
        return 'random'

    @property
    def max_questions(self):
        return 50

    def generate_scenario(self):
        dealer_card = random.randint(2, 11)
        hand_type = random.choice(['hard', 'soft', 'pair'])

        if hand_type == 'pair':
            pair_value = random.choice([2, 3, 4, 5, 6, 7, 8, 9, 10, 11])
            player_cards = [pair_value, pair_value]
            player_total = pair_value
        elif hand_type == 'soft':
            other_card = random.randint(2, 9)
            player_cards = [11, other_card]
            player_total = 11 + other_card
        else:  # hard
            player_total = random.randint(5, 20)
            player_cards = self._generate_hand_cards(hand_type, player_total)

        return hand_type, player_cards, player_total, dealer_card


class DealerGroupTrainingSession(TrainingSession):
    """Training session focused on specific dealer strength groups."""

    def __init__(self, difficulty='normal'):
        super().__init__(difficulty)
        self.dealer_group = None
        self._requires_submode = True

    @property
    def mode_name(self):
        return 'dealer_groups'

    @property
    def max_questions(self):
        return 50

    def setup_session(self):
        """Setup dealer group selection."""
        self.dealer_group = display_dealer_groups()
        return self.dealer_group

    def generate_scenario(self):
        # Select dealer card based on chosen group
        if self.dealer_group == 1:  # Weak
            dealer_card = random.choice([4, 5, 6])
        elif self.dealer_group == 2:  # Medium
            dealer_card = random.choice([2, 3, 7, 8])
        else:  # Strong
            dealer_card = random.choice([9, 10, 11])

        hand_type = random.choice(['hard', 'soft', 'pair'])

        if hand_type == 'pair':
            pair_value = random.choice([2, 3, 4, 5, 6, 7, 8, 9, 10, 11])
            player_cards = [pair_value, pair_value]
            player_total = pair_value
        elif hand_type == 'soft':
            other_card = random.randint(2, 9)
            player_cards = [11, other_card]
            player_total = 11 + other_card
        else:  # hard
            player_total = random.randint(5, 20)
            player_cards = self._generate_hand_cards(hand_type, player_total)

        return hand_type, player_cards, player_total, dealer_card


class HandTypeTrainingSession(TrainingSession):
    """Training session focused on specific hand types."""

    def __init__(self, difficulty='normal'):
        super().__init__(difficulty)
        self.hand_type_choice = None
        self._requires_submode = True

    @property
    def mode_name(self):
        return 'hand_types'

    @property
    def max_questions(self):
        return 50

    def setup_session(self):
        """Setup hand type selection."""
        self.hand_type_choice = display_hand_types()
        return self.hand_type_choice

    def generate_scenario(self):
        dealer_card = random.randint(2, 11)

        if self.hand_type_choice == 1:  # Hard totals
            player_total = random.randint(5, 20)
            player_cards = self._generate_hand_cards('hard', player_total)
            hand_type = 'hard'
        elif self.hand_type_choice == 2:  # Soft totals
            other_card = random.randint(2, 9)
            player_cards = [11, other_card]
            player_total = 11 + other_card
            hand_type = 'soft'
        else:  # Pairs
            pair_value = random.choice([2, 3, 4, 5, 6, 7, 8, 9, 10, 11])
            player_cards = [pair_value, pair_value]
            player_total = pair_value
            hand_type = 'pair'

        return hand_type, player_cards, player_total, dealer_card


class AbsoluteTrainingSession(TrainingSession):
    """Training session focused on absolute rules (always/never scenarios)."""

    @property
    def mode_name(self):
        return 'absolutes'

    @property
    def max_questions(self):
        return 20

    def generate_scenario(self):
        absolutes = [
            ('pair', [11, 11], 11),  # A,A
            ('pair', [8, 8], 8),     # 8,8
            ('pair', [10, 10], 10),  # 10,10
            ('pair', [5, 5], 5),     # 5,5
            ('hard', None, 17),      # Hard 17
            ('hard', None, 18),      # Hard 18
            ('hard', None, 19),      # Hard 19
            ('hard', None, 20),      # Hard 20
            ('soft', [11, 8], 19),   # Soft 19
            ('soft', [11, 9], 20),   # Soft 20
        ]

        hand_type, player_cards, player_total = random.choice(absolutes)
        dealer_card = random.randint(2, 11)

        if player_cards is None:  # Hard totals
            player_cards = self._generate_hand_cards(hand_type, player_total)

        return hand_type, player_cards, player_total, dealer_card

File: go/internal/trainer/trainer.go
// Package trainer provides training session types for blackjack strategy practice.
//
// This package defines the TrainingSession interface and implements concrete
// training session types that focus on different aspects of blackjack strategy:
// - RandomTrainingSession: Mixed practice with all hand types and dealer cards
// - DealerGroupTrainingSession: Focus on specific dealer strength groups
// - HandTypeTrainingSession: Focus on specific hand types (hard/soft/pairs)
// - AbsoluteTrainingSession: Practice absolute rules (always/never scenarios)
package trainer

import (
	"blackjack_trainer/internal/stats"
	"blackjack_trainer/internal/strategy"
	"blackjack_trainer/internal/ui"
	"fmt"
	"math/rand"
	"time"
)

// TrainingSession interface defines the contract for all training session types.
type TrainingSession interface {
	// GetModeName returns the mode name for display purposes.
	GetModeName() string
	// GetMaxQuestions returns the maximum number of questions for this session type.
	GetMaxQuestions() int
	// GenerateScenario generates a scenario for this training mode.
	// Returns (handType, playerCards, playerTotal, dealerCard).
	GenerateScenario() (strategy.HandType, []int, int, int)
	// SetupSession sets up the session. Returns true if setup successful, false if user cancelled.
	SetupSession() bool
}

// Scenario represents a training scenario.
type Scenario struct {
	HandType    strategy.HandType
	PlayerCards []int
	PlayerTotal int
	DealerCard  int
}

// BaseTrainer provides common functionality for all training sessions.
type BaseTrainer struct {
	rng *rand.Rand
}

// NewBaseTrainer creates a new base trainer with random number generator.
func NewBaseTrainer() *BaseTrainer {
	return &BaseTrainer{
		rng: rand.New(rand.NewSource(time.Now().UnixNano())),
	}
}

// GenerateHandCards generates card representation for a hand.
func (bt *BaseTrainer) GenerateHandCards(handType strategy.HandType, playerTotal int) []int {
	switch handType {
	case strategy.HandTypePair:
		return []int{playerTotal, playerTotal}
	case strategy.HandTypeSoft:
		otherCard := playerTotal - 11
		return []int{11, otherCard}
	case strategy.HandTypeHard:
		if playerTotal <= 11 {
			return []int{playerTotal}
		}
		// Generate two valid cards (2-10) that sum to playerTotal
		firstCard := bt.rng.Intn(min(9, playerTotal-2)) + 2
		secondCard := playerTotal - firstCard

		// If second card would be > 10, we need more cards
		if secondCard > 10 {
			// For totals > 20, generate 3+ cards
			cards := []int{firstCard}
			remaining := playerTotal - firstCard

			for remaining > 10 {
				// Take a card between 2 and min(10, remaining-2) to ensure we can finish
				maxCard := min(10, remaining-2)
				if maxCard < 2 {
					break
				}
				card := bt.rng.Intn(maxCard-1) + 2 // 2 to maxCard
				cards = append(cards, card)
				remaining -= card
			}

			if remaining >= 2 {
				cards = append(cards, remaining)
			}
			return cards
		} else if secondCard < 2 {
			// If second card would be < 2, just use single card
			return []int{playerTotal}
		} else {
			return []int{firstCard, secondCard}
		}
	default:
		return []int{playerTotal}
	}
}

// CheckAnswer checks if user's action matches the correct action.
func CheckAnswer(userAction, correctAction rune) bool {
	normalizedUser := userAction
	if userAction == 'P' {
		normalizedUser = 'Y'
	}
	return normalizedUser == correctAction
}

// RunSession runs the main training session loop.
func RunSession(session TrainingSession, statistics *stats.Statistics) {
	ui.DisplaySessionHeader(session.GetModeName())

	if !session.SetupSession() {
		return // User cancelled setup
	}

	strategyChart := strategy.New()
	var correctCount, totalCount, questionCount int

	for questionCount < session.GetMaxQuestions() {
		handType, playerCards, playerTotal, dealerCard := session.GenerateScenario()

		ui.DisplayHand(playerCards, dealerCard, handType, playerTotal)

		userAction, quit := ui.GetUserAction()
		if quit {
			break
		}

		correctAction := strategyChart.GetCorrectAction(handType, playerTotal, dealerCard)
		correct := CheckAnswer(userAction, correctAction)
		explanation := strategyChart.GetExplanation(handType, playerTotal, dealerCard)

		quitRequested := ui.DisplayFeedback(correct, userAction, correctAction, explanation)

		// Record statistics
		dealerStrength := statistics.GetDealerStrength(dealerCard)
		statistics.RecordAttempt(handType, dealerStrength, correct)

		questionCount++

		if correct {
			correctCount++
		}
		totalCount++

		if quitRequested {
			break
		}
	}

	// Show session summary
	if totalCount > 0 {
		accuracy := (float64(correctCount) / float64(totalCount)) * 100.0
		fmt.Printf("\nSession complete! Final score: %d/%d (%.1f%%)\n",
			correctCount, totalCount, accuracy)
	}
}

// RandomTrainingSession provides random practice with all hand types and dealer cards.
type RandomTrainingSession struct {
	*BaseTrainer
}

// NewRandomTrainingSession creates a new random training session.
func NewRandomTrainingSession() *RandomTrainingSession {
	return &RandomTrainingSession{
		BaseTrainer: NewBaseTrainer(),
	}
}

// GetModeName returns the mode name.
func (r *RandomTrainingSession) GetModeName() string {
	return "random"
}

// GetMaxQuestions returns the maximum number of questions.
func (r *RandomTrainingSession) GetMaxQuestions() int {
	return 50
}

// SetupSession sets up the session (no additional setup needed).
func (r *RandomTrainingSession) SetupSession() bool {
	return true
}

// GenerateScenario generates a random scenario.
func (r *RandomTrainingSession) GenerateScenario() (strategy.HandType, []int, int, int) {
	dealerCard := r.rng.Intn(10) + 2 // 2-11
	handTypes := []strategy.HandType{strategy.HandTypeHard, strategy.HandTypeSoft, strategy.HandTypePair}
	handType := handTypes[r.rng.Intn(len(handTypes))]

	var playerCards []int
	var playerTotal int

	switch handType {
	case strategy.HandTypePair:
		pairValues := []int{2, 3, 4, 5, 6, 7, 8, 9, 10, 11}
		pairValue := pairValues[r.rng.Intn(len(pairValues))]
		playerCards = []int{pairValue, pairValue}
		playerTotal = pairValue
	case strategy.HandTypeSoft:
		otherCard := r.rng.Intn(8) + 2 // 2-9
		playerCards = []int{11, otherCard}
		playerTotal = 11 + otherCard
	case strategy.HandTypeHard:
		playerTotal = r.rng.Intn(16) + 5 // 5-20
		playerCards = r.GenerateHandCards(strategy.HandTypeHard, playerTotal)
	}

	return handType, playerCards, playerTotal, dealerCard
}

// DealerGroupTrainingSession focuses on specific dealer strength groups.
type DealerGroupTrainingSession struct {
	*BaseTrainer
	dealerGroup int
}

// NewDealerGroupTrainingSession creates a new dealer group training session.
func NewDealerGroupTrainingSession() *DealerGroupTrainingSession {
	return &DealerGroupTrainingSession{
		BaseTrainer: NewBaseTrainer(),
		dealerGroup: 0,
	}
}

// GetModeName returns the mode name.
func (d *DealerGroupTrainingSession) GetModeName() string {
	return "dealer_groups"
}

// GetMaxQuestions returns the maximum number of questions.
func (d *DealerGroupTrainingSession) GetMaxQuestions() int {
	return 50
}

// SetupSession sets up the session by asking user to choose dealer group.
func (d *DealerGroupTrainingSession) SetupSession() bool {
	choice, ok := ui.DisplayDealerGroups()
	if !ok {
		return false
	}
	d.dealerGroup = choice
	return true
}

// GenerateScenario generates a scenario with specific dealer group.
func (d *DealerGroupTrainingSession) GenerateScenario() (strategy.HandType, []int, int, int) {
	// Select dealer card based on chosen group
	var dealerCard int
	switch d.dealerGroup {
	case 1: // Weak
		weakCards := []int{4, 5, 6}
		dealerCard = weakCards[d.rng.Intn(len(weakCards))]
	case 2: // Medium
		mediumCards := []int{2, 3, 7, 8}
		dealerCard = mediumCards[d.rng.Intn(len(mediumCards))]
	default: // Strong
		strongCards := []int{9, 10, 11}
		dealerCard = strongCards[d.rng.Intn(len(strongCards))]
	}

	handTypes := []strategy.HandType{strategy.HandTypeHard, strategy.HandTypeSoft, strategy.HandTypePair}
	handType := handTypes[d.rng.Intn(len(handTypes))]

	var playerCards []int
	var playerTotal int

	switch handType {
	case strategy.HandTypePair:
		pairValues := []int{2, 3, 4, 5, 6, 7, 8, 9, 10, 11}
		pairValue := pairValues[d.rng.Intn(len(pairValues))]
		playerCards = []int{pairValue, pairValue}
		playerTotal = pairValue
	case strategy.HandTypeSoft:
		otherCard := d.rng.Intn(8) + 2 // 2-9
		playerCards = []int{11, otherCard}
		playerTotal = 11 + otherCard
	case strategy.HandTypeHard:
		playerTotal = d.rng.Intn(16) + 5 // 5-20
		playerCards = d.GenerateHandCards(strategy.HandTypeHard, playerTotal)
	}

	return handType, playerCards, playerTotal, dealerCard
}

// HandTypeTrainingSession focuses on specific hand types.
type HandTypeTrainingSession struct {
	*BaseTrainer
	handTypeChoice int
}

// NewHandTypeTrainingSession creates a new hand type training session.
func NewHandTypeTrainingSession() *HandTypeTrainingSession {
	return &HandTypeTrainingSession{
		BaseTrainer:    NewBaseTrainer(),
		handTypeChoice: 0,
	}
}

// GetModeName returns the mode name.
func (h *HandTypeTrainingSession) GetModeName() string {
	return "hand_types"
}

// GetMaxQuestions returns the maximum number of questions.
func (h *HandTypeTrainingSession) GetMaxQuestions() int {
	return 50
}

// SetupSession sets up the session by asking user to choose hand type.
func (h *HandTypeTrainingSession) SetupSession() bool {
	choice, ok := ui.DisplayHandTypes()
	if !ok {
		return false
	}
	h.handTypeChoice = choice
	return true
}

// GenerateScenario generates a scenario with specific hand type.
func (h *HandTypeTrainingSession) GenerateScenario() (strategy.HandType, []int, int, int) {
	dealerCard := h.rng.Intn(10) + 2 // 2-11

	var handType strategy.HandType
	var playerCards []int
	var playerTotal int

	switch h.handTypeChoice {
	case 1: // Hard totals
		handType = strategy.HandTypeHard
		playerTotal = h.rng.Intn(16) + 5 // 5-20
		playerCards = h.GenerateHandCards(strategy.HandTypeHard, playerTotal)
	case 2: // Soft totals
		handType = strategy.HandTypeSoft
		otherCard := h.rng.Intn(8) + 2 // 2-9
		playerCards = []int{11, otherCard}
		playerTotal = 11 + otherCard
	default: // Pairs
		handType = strategy.HandTypePair
		pairValues := []int{2, 3, 4, 5, 6, 7, 8, 9, 10, 11}
		pairValue := pairValues[h.rng.Intn(len(pairValues))]
		playerCards = []int{pairValue, pairValue}
		playerTotal = pairValue
	}

	return handType, playerCards, playerTotal, dealerCard
}

// AbsoluteTrainingSession focuses on absolute rules (always/never scenarios).
type AbsoluteTrainingSession struct {
	*BaseTrainer
}

// NewAbsoluteTrainingSession creates a new absolute training session.
func NewAbsoluteTrainingSession() *AbsoluteTrainingSession {
	return &AbsoluteTrainingSession{
		BaseTrainer: NewBaseTrainer(),
	}
}

// GetModeName returns the mode name.
func (a *AbsoluteTrainingSession) GetModeName() string {
	return "absolutes"
}

// GetMaxQuestions returns the maximum number of questions.
func (a *AbsoluteTrainingSession) GetMaxQuestions() int {
	return 20
}

// SetupSession sets up the session (no additional setup needed).
func (a *AbsoluteTrainingSession) SetupSession() bool {
	return true
}

// GenerateScenario generates a scenario with absolute rules.
func (a *AbsoluteTrainingSession) GenerateScenario() (strategy.HandType, []int, int, int) {
	absolutes := []struct {
		handType    strategy.HandType
		playerCards []int
		playerTotal int
	}{
		{strategy.HandTypePair, []int{11, 11}, 11}, // A,A
		{strategy.HandTypePair, []int{8, 8}, 8},    // 8,8
		{strategy.HandTypePair, []int{10, 10}, 10}, // 10,10
		{strategy.HandTypePair, []int{5, 5}, 5},    // 5,5
		{strategy.HandTypeHard, []int{}, 17},       // Hard 17
		{strategy.HandTypeHard, []int{}, 18},       // Hard 18
		{strategy.HandTypeHard, []int{}, 19},       // Hard 19
		{strategy.HandTypeHard, []int{}, 20},       // Hard 20
		{strategy.HandTypeSoft, []int{11, 8}, 19},  // Soft 19
		{strategy.HandTypeSoft, []int{11, 9}, 20},  // Soft 20
	}

	absolute := absolutes[a.rng.Intn(len(absolutes))]
	dealerCard := a.rng.Intn(10) + 2 // 2-11

	playerCards := absolute.playerCards
	if len(playerCards) == 0 { // Hard totals
		playerCards = a.GenerateHandCards(absolute.handType, absolute.playerTotal)
	}

	return absolute.handType, playerCards, absolute.playerTotal, dealerCard
}

// Helper function to get minimum of two integers.
func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
